NitroxClient/GameLogic/Building.cs

```diff
@@ -1,213 +1,1326 @@
-            Optional<RotationMetadata> rotationMetadata = rotationMetadataFactory.From(baseGhost);
+            return true;
+        }
+
+
+        // Section: BuilderTool
+
+        public void BuilderTool_OnHoverConstructable_Post(GameObject gameObject, Constructable constructable)
+        {
+
+#if TRACE && BUILDING && HOVERCONSTRUCTABLE
+            NitroxId id = NitroxEntity.GetIdNullable(constructable.gameObject);
+            NitroxModel.Logger.Log.Debug("BuilderTool_OnHoverConstructable_Post - instance: " + constructable.gameObject.name + " id: " + id);
+#endif
 
-            BasePiece basePiece = new BasePiece(id, placedPosition, quaternion, camera.position, camera.rotation, techType.Model(), Optional.OfNullable(parentBaseId), false, rotationMetadata);
-            PlaceBasePiece placedBasePiece = new PlaceBasePiece(basePiece);
-            packetSender.Send(placedBasePiece);
+            lastHoveredConstructable = constructable;
         }
 
-        public void PlaceFurniture(GameObject gameObject, TechType techType, Vector3 itemPosition, Quaternion quaternion)
+        public void BuilderTool_OnHoverDeconstructable_Post(GameObject gameObject, BaseDeconstructable deconstructable)
         {
-            if (!Builder.isPlacing) //prevent possible echoing
+
+#if TRACE && BUILDING && HOVERDECONSTRUCTABLE
+            NitroxId id = NitroxEntity.GetIdNullable(deconstructable.gameObject);
+            NitroxId baseId = null;
+            Base abase = deconstructable.gameObject.GetComponentInParent<Base>();
+            if (abase)
             {
-                return;
+                baseId = NitroxEntity.GetIdNullable(abase.gameObject);
             }
+            NitroxModel.Logger.Log.Debug("BuilderTool_OnHoverDeconstructable_Post - instance: " + deconstructable.gameObject.name + " id: " + id + " baseId: " + baseId + " position: " + deconstructable.gameObject.transform.position + " rotation: " + deconstructable.gameObject.transform.rotation + " cellPosition: " + deconstructable.gameObject.transform.parent.position + " cellIndex: " + deconstructable.gameObject.transform.GetSiblingIndex());
+#endif
+
+        }
+
+        // Besides switching the state of currently handling a builderTool, this method is also intended to precheck if a construction/action even is allowed. If a 
+        // remote player is currently using a gameobject (e.g. Fabricator) or is too using the buildertool on the same object, we want to deny the local action here 
+        // and give a info to the player. 
+        public bool BuilderTool_HandleInput_Pre(GameObject gameObject)
+        {
 
-            NitroxId id = NitroxEntity.GetId(gameObject);
-            NitroxId parentId = null;
+#if TRACE && BUILDING && HOVER
+            NitroxModel.Logger.Log.Debug("BuilderTool_Pre_HandleInput");
+#endif
 
-            SubRoot sub = Player.main.currentSub;
+            currentlyHandlingBuilderTool = true;
+            return true;
```

Comment:
The return true is currently in any case true, because there is no implementation yet wether the targetted object is in use by another player or not. When this is implemented, the return will be false in this case. This will suppress to execute the entire handleinput function of subnautica. An forbidden info will be given to the user as a visual (e.g. deconstruction not allowed in case of faces attached).

Reply:
Sorry I don't think I was fully clear.  The patch should make the determination if it should suppress the code that is being mutated.  This is because the patch is close to the implementation and understands what the upcoming code will be doing.  The code in gamelogic is generally agnostic to who is calling it and sometimes has multiple invokers (such as through patches or packet processors).  With that being said, the patches can leverage gamelogic code to make an informed decision but we should tie it behind a descriptive name such as isRemotePlayerAlreadyInteracting().  I think any time that you are invoking a gamelogic method from a patch you should tie it to a descriptive name.